keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/ImageService.kt

```diff
@@ -83,110 +83,91 @@ class ImageService(
+      cloudDriverService.namedImages(
+        user = DEFAULT_SERVICE_ACCOUNT,
+        imageName = appVersion.toImageName().replace("~", "_"),
+        account = account
+      )
+        // only consider images with tags and app version set properly
+        .asSequence()
+        .filter { image ->
+          tagsExistForAllAmis(image.tagsByImageId) && image.hasAppVersion
         }
-      }
+        // filter to images with matching app version
+        .filter { image ->
+          // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
+          runCatching { AppVersion.parseName(image.appVersion) }
+            .getOrElse { ex ->
+              throw SystemException("trying to parse name for image ${image.imageName} with version ${image.appVersion} but got an exception", ex)
+            }
+            .run {
+              packageName == appVersion.packageName && version == appVersion.version && commit == appVersion.commit
+            }
+        }
+        // filter to images in the correct account and the desired region
+        .filter { image ->
+          image.accounts.contains(account) && image.amis.containsKey(region)
+        }
+        // reduce to the newest images required to support all regions we want
+        .sortedByDescending { it.creationDate }
+        .firstOrNull()
+  }
 
   /**
-   * Returns the latest image that is present in all regions.
-   * Each ami must have tags.
+   * Find the latest properly tagged image in [account] and [region].
+   *
+   * As a side effect this method will prime the cache for any additional regions where the image is
+   * available.
    */
-  suspend fun getLatestNamedImageWithAllRegionsForAppVersion(appVersion: AppVersion, account: String, regions: Collection<String>): NamedImage? =
-    cloudDriverService.namedImages(
-      user = DEFAULT_SERVICE_ACCOUNT,
-      imageName = appVersion.toImageName().replace("~", "_"),
-      account = account
-    )
-      .filter { it.hasAppVersion }
-      .sortedWith(NamedImageComparator)
-      .find { namedImage ->
-        // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
-        val curAppVersion = try {
-          AppVersion.parseName(namedImage.appVersion)
-        } catch (ex: Exception) {
-          throw SystemException("trying to parse name for image ${namedImage.imageName} with version ${namedImage.appVersion} but got an exception", ex)
+  suspend fun getLatestNamedImageForAppVersionInRegion(
+    appVersion: AppVersion,
+    account: String,
+    region: String
+  ): NamedImage? =
+    namedImageCache
+      .get(NamedImageCacheKey(appVersion, account, region))
+      .await()
+      // prime the cache if the image is also in other regions
+      ?.also { image ->
+        (image.amis.keys - region).forEach { otherRegion ->
+          namedImageCache.put(NamedImageCacheKey(appVersion, account, otherRegion), completedFuture(image))
         }
```

Comment:
Method to find a single region just refers to the cache that will load on demand from CloudDriver. If the resulting image actually supports other regions besides the desired one, it pre-seeds the cache for those other regions too which will make future lookups (or lookups happening in parallel) more efficient.

Reply:
Nice! ❤️